### PR TITLE
properly handle files with uri-encoded components

### PR DESCRIPTION
**Pull request by timcharper**
vfs.resolveFile receives a String URI, and will parse and decode URI
components. This means if the file name contains URI-encoded components
(such as %3A, as courier will output to escape unsafe FS characters like
:), they will be decoded, and the wrong filename will be received.

fixes https://github.com/ensime/ensime-server/issues/1457

### CODE REVIEW

**Comment by timcharper** [comment]
Sorry so sloppy, I didn't write tests for this.

**Comment by timcharper** [comment]
Okay there is a test now. It's not the best test, it's a test and it beats all the rest.

**Comment by timcharper** [comment]
The other tests in `util/src/test/scala/org/ensime/util` aren't using the mentioned `EnsimeSpec`. It appears that the class isn't available.

**Comment by fommil** [comment]
oh right, `testutil` depends on `utils`

File: util/src/main/scala/org/ensime/vfs/EnsimeVFS.scala [hunk 1]
```diff
@@ -46,6 +46,20 @@ object `package` {
   private val zipFileField = classOf[ZipFileSystem].getDeclaredField("zipFile")
   zipFileField.setAccessible(true)
 
+  object URLParamEncoder {
+    private val unsafe = " %"
+    def encode(input: String): String = {
+      val resultStr = new StringBuilder()
+      input.toCharArray.foreach { ch =>
+        if (unsafe contains ch)
```
Comment (fommil) [inline util/src/main/scala/org/ensime/vfs/EnsimeVFS.scala]: I shudder to think how inefficient this is in terms of scala collections.

**Comment by fommil** [comment]
LGTM merge when you're happy

**Comment by fommil** [comment]
oh, looks like you need to rebase before you can do that

File: util/src/main/scala/org/ensime/vfs/EnsimeVFS.scala (hunk 1 above)
Comment (timcharper) [inline util/src/main/scala/org/ensime/vfs/EnsimeVFS.scala]: good point; I'll use a set instead

Comment (fommil) [inline util/src/main/scala/org/ensime/vfs/EnsimeVFS.scala]: I'd just use an `if` statement to be honest. The object creation that happens when `String` is treated as a `Traversable` is pretty crazy.

File: util/src/main/scala/org/ensime/vfs/EnsimeVFS.scala [hunk 2]
```diff
@@ -46,6 +46,22 @@ object `package` {
   private val zipFileField = classOf[ZipFileSystem].getDeclaredField("zipFile")
   zipFileField.setAccessible(true)
 
+  object URLParamEncoder {
+    def encode(input: String): String = {
```
Comment (lukasz-golebiewski) [inline util/src/main/scala/org/ensime/vfs/EnsimeVFS.scala]: Won't `input.map{ c =>  if ((c == ' ') || (c == '%')) "%%%02x".format(c.toShort) else c }` do the same thing?

**Comment by fommil** [comment]
heh, sorry, just merged another PR. Seems github is getting really fussy about that nowadays! It used to be happy with PRs that had a clean rebase.

File: util/src/main/scala/org/ensime/vfs/EnsimeVFS.scala (hunk 2 above)
Comment (timcharper) [inline util/src/main/scala/org/ensime/vfs/EnsimeVFS.scala]: Yes, but not as efficiently. I think?

Comment (timcharper) [inline util/src/main/scala/org/ensime/vfs/EnsimeVFS.scala]: trying to reduce allocations since this is a bit of a hot-path in this code

Comment (fommil) [inline util/src/main/scala/org/ensime/vfs/EnsimeVFS.scala]: isn't that returning `String` or `Char`? From an efficiency point of view we create a lot of `URI`s so the current approach could contribute to quite a bit of churn. To really tackle that we might want to drop the `foreach` and use a `while` loop. But no need to prematurely optimise, let's see if it's a problem.

Comment (timcharper) [inline util/src/main/scala/org/ensime/vfs/EnsimeVFS.scala]:
```
scala> "asdf".map { _ => "B" }
res0: scala.collection.immutable.IndexedSeq[String] = Vector(B, B, B, B)
```

Yeah... definitely not as efficient.

Comment (timcharper) [inline util/src/main/scala/org/ensime/vfs/EnsimeVFS.scala]: @fommil I've already dropped the `foreach`

**Comment by fommil** [comment]
@rorygraves will be very proud :grinning: https://skillsmatter.com/skillscasts/6939-optimising-scala-for-fun-and-profit

**Comment by fommil** [comment]
I really wish we had parallel matrix builds in drone...

File: util/src/main/scala/org/ensime/vfs/EnsimeVFS.scala (hunk 2 above)
Comment (lukasz-golebiewski) [inline util/src/main/scala/org/ensime/vfs/EnsimeVFS.scala]: What about `input.map{ c => if ((c == ' ') || (c == '%')) "%%%02x".format(c.toShort) else c }(scala.collection.breakOut)` ? ;-)

Comment (lukasz-golebiewski) [inline util/src/main/scala/org/ensime/vfs/EnsimeVFS.scala]: LGTM! Didn't notice you were mapping to a `String`

Comment (lukasz-golebiewski) [inline util/src/main/scala/org/ensime/vfs/EnsimeVFS.scala]: You could always do 
```
@ "asdb".flatMap(_ => "B")(scala.collection.breakOut)
res2: String = "BBBB" 
```

Comment (fommil) [inline util/src/main/scala/org/ensime/vfs/EnsimeVFS.scala]: have you done an object allocation profile of that? I bet it's not pretty! :laughing:

**Comment by fommil** [comment]
you managed to get out of date with master again

**Comment by fommil** [comment]
ok, I removed the (new) restriction that branches have to be up to date. It might bite us occasionally, but I'm willing to live dangerously for the convenience.

**Comment by fommil** [comment]
oh it's bedtime in your TZ, I'll merge.

File: util/src/main/scala/org/ensime/vfs/EnsimeVFS.scala (hunk 2 above)
Comment (lukasz-golebiewski) [inline util/src/main/scala/org/ensime/vfs/EnsimeVFS.scala]: You're right :) `for (x <- this) b ++= f(x).seq`

**Comment by rorygraves** [comment]
@fommil  - Proud, but sad that this kind of stuff is still required.

**Comment by fommil** [comment]
it would be interesting to use one of those automagic cleanup plugins you talked about, see if it makes a difference on a huge project.

**Pull request merged**